ble-shelly-btn.js

```diff
@@ -0,0 +1,221 @@
+    if (type === int8) return this.getInt8();
+    if (type === uint16) return this.getUInt16LE();
+    if (type === int16) return this.getInt16LE();
+    if (type === uint24) return this.getUInt24LE();
+    if (type === int24) return this.getInt24LE();
+    return null;
+  },
+  unpack: function () {
+    if (this.buffer === null) return null;
+    let result = {};
+    let dib = this.buffer.at(0);
+    result["encryption"] = dib & 0x1 ? true : false;
+    result["BTHome_version"] = dib >> 5;
+    if (result["BTHome_version"] !== 2) return null;
+    //Can not handle encrypted data
+    if (result["encryption"] === 1) return null;
+    this.buffer = this.buffer.slice(1);
+
+    while (this.buffer.length > 0) {
+      let _bth = BTH[this.buffer.at(0)];
+      if (_bth === "undefined") return null;
+      this.buffer = this.buffer.slice(1);
+      let _value = this.getBufValue(_bth.t);
+      if (_value === null) return null;
+      if (typeof _bth.f !== "undefined") _value = _value * _bth.f;
+      result[_bth.n] = _value;
+      this.buffer = this.buffer.slice(getByteSize(_bth.t));
+    }
+    return result;
+  },
+};
+
+let ShellyBLUParser = {
+  getData: function (res) {
+    BTHomeDecoder.setBuffer(res.service_data[BTHOME_SVC_ID_STR]);
+    let result = BTHomeDecoder.unpack();
+    return result;
+  },
+};
+
+let last_pid = 0x100;
+function scanCB(ev, res) {
+  if (ev !== BLE.Scanner.SCAN_RESULT) return;
+  // skip if there is no service_data member
+  if (
+    typeof res.service_data === "undefined" ||
+    typeof res.service_data[BTHOME_SVC_ID_STR] === "undefined"
+  )
+    return;
+  // skip if we are looking for name match but don't have active scan as we don't have name
+  if (
+    typeof CONFIG.shelly_blu_name_prefix !== "undefined" &&
+    (typeof res.local_name === "undefined" ||
+      res.local_name.indexOf(CONFIG.shelly_blu_name_prefix) !== 0)
+  )
+    return;
+  // skip if we don't have address match
+  if (
+    typeof CONFIG.shelly_blu_address !== "undefined" &&
+    CONFIG.shelly_blu_address !== res.addr
+  )
+    return;
+  let BTHparsed = ShellyBLUParser.getData(res);
+  // skip if parsing failed
+  if (BTHparsed === null) return;
+  // skip, we are deduping results
+  if (last_pid === BTHparsed.pid) return;
+  last_pid = BTHparsed.pid;
+  console.log(JSON.stringify(BTHparsed));
+  let BLEData = {
+    addr: res.addr,
+    rssi: res.rssi,
+  };
+  // execute actions from CONFIG
+  let aIdx = null;
+  for (aIdx in CONFIG.actions) {
+    // skip if no conditionn defined
+    if (typeof CONFIG.actions[aIdx]["cond"] === "undefined") continue;
+    let cond = CONFIG.actions[aIdx]["cond"];
+    let cIdx = null;
+    let run = true;
+    for (cIdx in cond) {
+      if (typeof BTHparsed[cIdx] === "undefined") run = false;
+      if (BTHparsed[cIdx] !== cond[cIdx]) run = false;
+    }
+    // if all conditions evaluated to true then execute
+    if (run) CONFIG.actions[aIdx]["action"](BTHparsed, BLEData);
```

Comment:
If BTHparsed, BLEData are combined in a single BTMessage object as suggested above actions can be simplified to take a single argument.